python/src/iceberg/catalog/base.py

```diff
@@ -0,0 +1,198 @@
+    ) -> Table:
+        """
+        Create a table
+
+        :param name: Table's name. Fully classified table name, if it is a namespaced catalog.
+        :param schema: Table's schema
+        :param partition_spec: A partition spec for the table
+        :param location: a location for the table; Optional Keyword Argument
+        :param properties: a string dictionary of table properties; Optional Keyword Argument
+        :return: the created table instance
+        :raises AlreadyExistsError: If a table with the name already exists
+        """
+
+    @abstractmethod
+    def table(self, name: str) -> Table:
+        """
+        Loads the table's metadata and returns the table instance. You can also use this method to
+        check for table existence using 'try catalog.table() except TableNotFoundError'
+        Note: This method does not load table's data in any form.
+
+        :param name: Table's name. Fully classified table name, if it is a namespaced catalog.
+        :return: the table instance with its metadata
+        :raises TableNotFoundError: If a table with the name does not exist
+        """
+
+    @abstractmethod
+    def drop_table(self, name: str, purge: bool = True) -> None:
+        """
+        Drop a table; Optionally purge all data and metadata files.
+
+        :param name: table name
+        :param purge: Defaults to true, which deletes all data and metadata files in the table; Optional Argument
+        :return: Nothing
+        :raises TableNotFoundError: If a table with the name does not exist
+        """
+
+    @abstractmethod
+    def rename_table(self, from_name: str, to_name: str) -> None:
+        """
+        Drop a table; Optionally purge all data and metadata files.
+
+        :param from_name: Existing table's name. Fully classified table name, if it is a namespaced catalog.
+        :param to_name: New Table name to be assigned. Fully classified table name, if it is a namespaced catalog.
+        :return: Nothing
+        :raises TableNotFoundError: If a table with the name does not exist
+        """
+
+    @abstractmethod
+    def replace_table(
+        self,
+        name: str,
+        schema: Schema,
+        partition_spec: PartitionSpec,
+        *,
+        location: Optional[str] = None,
+        properties: Optional[dict] = None
+    ) -> Table:
+        """
+        Starts a transaction and replaces the table with the provided spec.
+
+        :param name: Table's name. Fully classified table name, if it is a namespaced catalog.
+        :param schema: Table's schema
+        :param partition_spec: A partition spec for the table
+        :param location: a location for the table; Optional Keyword Argument
+        :param properties: a string dictionary of table properties; Optional Keyword Argument
+        :return: the replaced table instance
+        :raises TableNotFoundError: If a table with the name does not exist
+        """
+
+
+class NamespacedCatalog(Catalog):
```

Comment:
Do we need to have this as a subclass? I know the JDBC catalog has been updated recently to include namespaces. If the ideal implementation of a catalog should support namespaces, then I'd vote for adding these abstractmethods to `Catalog(ABC)` and an implementation that does not support namespaces will have to explicitly override it and raise some kind of error.

Reply:
Tell me about it :) I was torn on this one as well as I too strongly felt that `Catalog(ABC)` and `NamespacedCatalog(Catalog)` could be just one. I could not find even a single catalog implementation in the Java API that did NOT support namespaces. All of them do, so it seemed like a redundant extensibility being offered.

Reply:
Added all the namespace methods to just `Catalog(ABC)`. Resolving this.